### PR TITLE
state: node bootstrap in v0.37.x

### LINKED ISSUE #1513: Backport state bootstrap API to v0.37.x
### Summary

The original description of the problem and the need for is can be found in issue #884 and ADR 101 from #801.

### Problem Definition

SDK now allows their users to perform statesync offline and boot up the node afterwards. At the moment, the stores are bootstrapped with code within the SDK which was copied from CometBFT's statesync. However, there are some concerns with whether this leaves the stores in a correct state. Furthermore, they should not be required to maintain and fix bugs in code that should utlimately be done by Comet. 
As there are already users using this feature, we want to roll out a API for them to use ASAP to avoid any problems with offline state sync.

This needs to be done in a non-breaking way (including non API breaking).

### Proposal

Backport the PR https://github.com/cometbft/cometbft/pull/1057

### PR DESCRIPTION
**Pull request by jmalicevic**
Manual backport of #1057 
Closes #1513 

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: state/store.go [hunk 1]
```diff
@@ -75,6 +77,10 @@ type Store interface {
 	PruneStates(int64, int64) error
 	// Close closes the connection with the database
 	Close() error
+	// // Saves the height at which the store is bootstrapped after out of band statesync
+	// SetOfflineStateSyncHeight(height int64) error
+	// // Gets the height at which the store is bootstrapped after out of band statesync
+	// GetOfflineStateSyncHeight() (int64, error)
```
Comment (sergio-mena) [inline state/store.go:83]: Do we need this?

File: consensus/state.go [hunk 2]
```diff
@@ -171,25 +174,31 @@ func NewState(
...
 	cs.setProposal = cs.defaultSetProposal
 
 	// We have no votes, so reconstruct LastCommit from SeenCommit.
 	if state.LastBlockHeight > 0 {
+		// In case of out of band performed statesync, the state store
+		// will have a state but no extended commit (as no block has been downloaded).
+		// If the height at which the vote extensions are enabled is lower
+		// than the height at which we statesync, consensus will panic because
+		// it will try to reconstruct the extended commit here.
+
```
Comment (sergio-mena) [inline consensus/state.go:192]:
```suggestion
```

File: consensus/state.go [hunk 3]
```diff
@@ -171,25 +174,31 @@ func NewState(
...
 		cs.reconstructLastCommit(state)
+
```
Comment (sergio-mena) [inline consensus/state.go:194]:
```suggestion
```

File: state/store.go [hunk 4]
```diff
@@ -658,6 +672,56 @@ func (store dbStore) saveConsensusParamsInfo(nextHeight, changeHeight int64, par
 	return nil
 }
 
+/* Custom struct to handle offlien state sync
```
Comment (sergio-mena) [inline state/store.go:675]:
```suggestion
/* Custom struct to handle offline state sync
```

File: blocksync/reactor.go [hunk 5]
```diff
@@ -60,21 +60,34 @@ type Reactor struct {
...
-// NewReactor returns new reactor instance.
-func NewReactor(state sm.State, blockExec *sm.BlockExecutor, store *store.BlockStore,
-	blockSync bool) *Reactor {
+func NewReactorWithOfflineStateSync(state sm.State, blockExec *sm.BlockExecutor, store *store.BlockStore,
+	blockSync bool, offlineStateSyncHeight int64) *Reactor {
 
 	if state.LastBlockHeight != store.Height() {
 		panic(fmt.Sprintf("state (%v) and store (%v) height mismatch", state.LastBlockHeight,
 			store.Height()))
 	}
```
Comment (sergio-mena) [inline blocksync/reactor.go:70]: Remove? (from diff-of-diffs)

File: state/store.go (hunk 1 above)
Comment (jmalicevic) [reply state/store.go]: removed

**Review by sergio-mena — APPROVED** [review]
LGTM, thanks for this

**Pull request merged**